From 792a545bab3ade72830240be0e472b7bc6da3c8e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Marcin=20Wcis=C5=82o?= <marcin.wcislo@conclusive.pl>
Date: Thu, 5 Feb 2026 21:08:22 +0100
Subject: [PATCH] mm: zswap: fix missing folio cleanup in writeback race path
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

jira VULN-175655
cve CVE-2024-26832
commit-author Yosry Ahmed <yosryahmed@google.com>
commit e3b63e966cac0bf78aaa1efede1827a252815a1d
upstream-diff Manual change; no commit was actually cherry-picked as the
  auto resolver puts the change completely out of place and its contents
  had to be rewritten entirely anyway. Kernel ciqlts8_6 lacks the
  transition from page to folio in `zswap_writeback_entry()' introduced
  in 96c7b0b42239e7b8987b2664b458dc74e825f760, so in the backported
  version `unlock_page(page)' is used instead of `folio_unlock(folio)'
  and `put_page(page)' instead of `folio_put(folio)'. (See also the
  relevant, non-backported commit
  4e1364286d0a2dd384bceb6db6185b99c0e2c0bc). In the upstream, up until
  e3b63e9, the `zswap_writeback_entry()' function underwent major
  refactors in ff9d5ba202f98db53da98330b46b68a9228c63e4,
  98804a944a63237814257dd149a5b04d6b93489c,
  32acba4c04830487ca3002d716325e02069e053a and 96c7b0b compared to the
  ciqlts8_6 version, but the writeback race path addressed by the fix
  remained largely intact, and in ciqlts8_6 falls under the
  ZSWAP_SWAPCACHE_NEW case.

In zswap_writeback_entry(), after we get a folio from
__read_swap_cache_async(), we grab the tree lock again to check that the
swap entry was not invalidated and recycled.  If it was, we delete the
folio we just added to the swap cache and exit.

However, __read_swap_cache_async() returns the folio locked when it is
newly allocated, which is always true for this path, and the folio is
ref'd.  Make sure to unlock and put the folio before returning.

This was discovered by code inspection, probably because this path handles
a race condition that should not happen often, and the bug would not crash
the system, it will only strand the folio indefinitely.

Link: https://lkml.kernel.org/r/20240125085127.1327013-1-yosryahmed@google.com
Fixes: 04fc7816089c ("mm: fix zswap writeback race condition")
	Signed-off-by: Yosry Ahmed <yosryahmed@google.com>
	Reviewed-by: Chengming Zhou <zhouchengming@bytedance.com>
	Acked-by: Johannes Weiner <hannes@cmpxchg.org>
	Reviewed-by: Nhat Pham <nphamcs@gmail.com>
	Cc: Domenico Cerasuolo <cerasuolodomenico@gmail.com>
	Cc: <stable@vger.kernel.org>
	Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
(cherry picked from commit e3b63e966cac0bf78aaa1efede1827a252815a1d)
	Signed-off-by: Marcin Wcisło <marcin.wcislo@conclusive.pl>
---
 mm/zswap.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/mm/zswap.c b/mm/zswap.c
index 4c0b81b00f5c..0dfbdfd00757 100644
--- a/mm/zswap.c
+++ b/mm/zswap.c
@@ -904,6 +904,8 @@ static int zswap_writeback_entry(struct zpool *pool, unsigned long handle)
 		if (zswap_rb_search(&tree->rbroot, entry->offset) != entry) {
 			spin_unlock(&tree->lock);
 			delete_from_swap_cache(page);
+			unlock_page(page);
+			put_page(page);
 			ret = -ENOMEM;
 			goto fail;
 		}
-- 
2.47.3

